File: src/test/java/niklasu/speedtester/config/ParamValidatorTest.java
package niklasu.speedtester.config;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.io.IOException;
import java.net.URL;

import org.junit.jupiter.api.Test;

class ParamValidatorTest {

    @Test
    void NetworkError() throws IOException, FileSizeChecker.NoFileSizeException, ValidationException {
        URL url = new URL("http://fred.de");
        FileSizeChecker fileSizeChecker = mock(FileSizeChecker.class);
        when(fileSizeChecker.getFileSize(any())).thenThrow(new IOException("kaputt"));

        ParamValidator paramValidator = new ParamValidator(fileSizeChecker);

        assertThrows(ValidationException.class, () -> paramValidator.validate(100L, url.toString()));
    }

    @Test
    void malformedURL() {
        FileSizeChecker fileSizeChecker = mock(FileSizeChecker.class);
        ParamValidator paramValidator = new ParamValidator(fileSizeChecker);

        assertThrows(ValidationException.class, () -> paramValidator.validate(100L, "httasdddasdp://fred.de"));

    }

    @Test
    void fileIsTooSmall() throws IOException, FileSizeChecker.NoFileSizeException, ValidationException {
        FileSizeChecker fileSizeChecker = mock(FileSizeChecker.class);
        when(fileSizeChecker.getFileSize(any())).thenReturn(1L);
        ParamValidator paramValidator = new ParamValidator(fileSizeChecker);
        assertThrows(ValidationException.class, () -> paramValidator.validate(100L, "http://fred.de"));
    }

}

File: src/test/java/niklasu/speedtester/config/ParamValidatorTest.kt
package niklasu.speedtester.config

import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import java.io.IOException
import java.net.URL

internal class ParamValidatorTest {

    @Test
    fun NetworkError() {
        val url = URL("http://fred.de")
        val fileSizeChecker = mockk<FileSizeChecker> { every { getFileSize(any()) } throws IOException("kaputt") }
        val paramValidator = ParamValidator(fileSizeChecker)

        assertThrows(ValidationException::class.java) { paramValidator.validate(100, 2, url.toString()) }.printStackTrace()
    }

    @Test
    @Throws(java.io.IOException::class)
    fun malformedURL() {
        val fileSizeChecker = mockk<FileSizeChecker> {}
        val paramValidator = ParamValidator(fileSizeChecker)


        assertThrows(ValidationException::class.java) { paramValidator.validate(100, 2, "httasdddasdp://fred.de") }.printStackTrace()
    }

    @Test
    @Throws(java.io.IOException::class)
    fun fileIsTooSmall() {
        val url = URL("http://fred.de")
        val fileSizeChecker = mockk<FileSizeChecker> { every { getFileSize(url) } returns (1) }
        val paramValidator = ParamValidator(fileSizeChecker)
        assertThrows(ValidationException::class.java) { paramValidator.validate(100, 2, "http://fred.de") }.printStackTrace()
    }

    @Test
    @Throws(java.io.IOException::class)
    fun intervalLessThanOne() {
        val url = URL("http://fred.de")
        val fileSizeChecker = mockk<FileSizeChecker> { every { getFileSize(url) } returns (99999999999999) }
        val paramValidator = ParamValidator(fileSizeChecker)

        assertThrows(ValidationException::class.java) { paramValidator.validate(100, 0, "http://fred.de") }.printStackTrace()
    }
}